cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -830,38 +821,51 @@ Status AppendPySequence(PyObject* obj, int64_t size,
   return converter->AppendData(obj, size);
 }
 
-Status ConvertPySequence(PyObject* obj, MemoryPool* pool, std::shared_ptr<Array>* out) {
+static Status ConvertPySequenceReal(PyObject* obj, MemoryPool* pool,
+                                    std::shared_ptr<Array>* out,
+                                    const std::shared_ptr<DataType>* type, int64_t size) {
   PyAcquireGIL lock;
-  std::shared_ptr<DataType> type;
-  int64_t size;
-  RETURN_NOT_OK(InferArrowTypeAndSize(obj, &size, &type));
-  return ConvertPySequence(obj, pool, out, type, size);
-}
+  ScopedRef seqref;
+  std::shared_ptr<DataType> real_type;
+
+  RETURN_NOT_OK(ConvertToSequenceAndInferSize(obj, seqref.ref(), &size));
```

Comment:
I think `seqref.ref()` is only writable on account of `const PyObject*` being unwieldy -- it might be less awkward to pass a mutable pointer to `seqref`